### PR TITLE
refactor: Made PendingEventsStore interface asynchronous

### PR DESCRIPTION
**Pull request by zashraf1985**
## Summary
Made PendingEventsStore interface asynchronous. This change is done to implement React Native Async Storage which is asynchronous using the same interface.

## Test plan
Fixed the failing unit tests

### CODE REVIEW

File: packages/event-processor/__tests__/pendingEventsDispatcher.spec.ts [hunk 1]
```diff
@@ -61,22 +61,25 @@ describe('LocalStoragePendingEventsDispatcher', () => {
...
-    internalDispatchCall[1]({ statusCode: 200 })
-
-    // assert that the original dispatch function was called with the request
-    expect((originalEventDispatcher.dispatchEvent as unknown) as jest.Mock).toBeCalledTimes(1)
-    expect(internalDispatchCall[0]).toEqual(eventV1Request)
-
-    // assert that the passed in callback to pendingEventsDispatcher was called
-    expect(callback).toHaveBeenCalledTimes(1)
-    expect(callback).toHaveBeenCalledWith({ statusCode: 200 })
+    setTimeout(() => {
```
Comment (mjc1283) [inline packages/event-processor/__tests__/pendingEventsDispatcher.spec.ts:64]: Why did you add this `setTimeout`?

File: packages/event-processor/src/pendingEventsDispatcher.ts [hunk 2]
```diff
@@ -53,23 +53,22 @@ export class PendingEventsDispatcher implements EventDispatcher {
...
   sendPendingEvents(): void {
-    const pendingEvents = this.store.values()
-
-    logger.debug('Sending %s pending events from previous page', pendingEvents.length)
-
-    pendingEvents.forEach(item => {
-      try {
-        this.send(item, () => {})
-      } catch (e) {}
+    this.store.values().then((pendingEvents) => {
```
Comment (mjc1283) [inline packages/event-processor/src/pendingEventsDispatcher.ts:56]: - Did you test this in the browser to see how its behavior compares to the old version? I'm curious if the additional async step would affect how much of this is allowed by the browser to execute during a page close.
- If `close` is called while `values()` is in progress, I think we need to stop and not send the events. To the extend possible, we should stop all additional work after `close` is called.

File: packages/event-processor/src/pendingEventsDispatcher.ts [hunk 3]
```diff
@@ -53,23 +53,22 @@ export class PendingEventsDispatcher implements EventDispatcher {
...
     })
   }
 
   protected send(entry: DispatcherEntry, callback: EventDispatcherCallback): void {
-    this.store.set(entry.uuid, entry)
-
-    this.dispatcher.dispatchEvent(entry.request, response => {
-      this.store.remove(entry.uuid)
-      callback(response)
+    this.store.set(entry.uuid, entry).then(() => {
```
Comment (mjc1283) [inline packages/event-processor/src/pendingEventsDispatcher.ts:67]: Same comment as above RE: doing things after close: If `close` is called while `set(...)` is in progress, we need to stop and not send the events.

File: packages/event-processor/src/pendingEventsDispatcher.ts [hunk 4]
```diff
@@ -53,23 +53,22 @@ export class PendingEventsDispatcher implements EventDispatcher {
...
+      this.dispatcher.dispatchEvent(entry.request, response => {
+        this.store.remove(entry.uuid)
```
Comment (mjc1283) [inline packages/event-processor/src/pendingEventsDispatcher.ts:69]: I think we need to wait for remove to complete before calling the callback. This callback eventually feeds into the `close` promise status. We don't want `close` to resolve while an async operation is still pending.

File: packages/event-processor/src/pendingEventsDispatcher.ts (hunk 3 above)
Comment (mjc1283) [reply packages/event-processor/src/pendingEventsDispatcher.ts:67]: Discussed this offline w/Zeeshan - this part is OK as-is, disregard.

File: packages/event-processor/__tests__/pendingEventsDispatcher.spec.ts (hunk 1 above)
Comment (zashraf1985) [reply packages/event-processor/__tests__/pendingEventsDispatcher.spec.ts:64]: Because i want to detect call `dispatchEvent` on the original event dispatcher which was called synchronously earlier but now its being called asynchronously after the `set` promise resolves.

File: packages/event-processor/src/pendingEventsDispatcher.ts (hunk 4 above)
Comment (zashraf1985) [reply packages/event-processor/src/pendingEventsDispatcher.ts]: Done

File: packages/event-processor/src/pendingEventsDispatcher.ts [hunk 5]
```diff
@@ -53,23 +53,21 @@ export class PendingEventsDispatcher implements EventDispatcher {
...
-
-    pendingEvents.forEach(item => {
-      try {
-        this.send(item, () => {})
-      } catch (e) {}
+    this.store.values().then((pendingEvents) => {
+      logger.debug('Sending %s pending events from previous page', pendingEvents.length)
+      pendingEvents.forEach(item => {
+        try {
+          this.send(item, () => {})
```
Comment (jaeopt) [inline packages/event-processor/src/pendingEventsDispatcher.ts:60]: Does MAP guarantee the fifo order in JS? Otherwise, we lose the order of events dispatched.

Comment (jaeopt) [reply packages/event-processor/src/pendingEventsDispatcher.ts:60]: Also, orders are concerned in other cases as well:
- if we have any queued events, all the following events dispatched should be queued as well instead of sending them first. We may have queued events from the previous session or new events queued from the last failed dispatch.

File: packages/event-processor/src/pendingEventsDispatcher.ts [hunk 6]
```diff
@@ -53,23 +53,21 @@ export class PendingEventsDispatcher implements EventDispatcher {
...
 
   protected send(entry: DispatcherEntry, callback: EventDispatcherCallback): void {
-    this.store.set(entry.uuid, entry)
-
-    this.dispatcher.dispatchEvent(entry.request, response => {
-      this.store.remove(entry.uuid)
-      callback(response)
+    this.store.set(entry.uuid, entry).then(() => {
+      this.dispatcher.dispatchEvent(entry.request, response => {
+        this.store.remove(entry.uuid).then(() => callback(response))
```
Comment (jaeopt) [inline packages/event-processor/src/pendingEventsDispatcher.ts:69]: We also need a protection from one packet stuck in the queue forever (cannot send to the server successfully for JSON format error etc). This can be an issue when we change them in FIFO order.

**Comment by zashraf1985** [comment]
This is no longer needed now. I am now leaving the existing localstorge implementation intact and implementing event caching for react native separately in a new PR.
https://github.com/optimizely/javascript-sdk/pull/517
